src/org/etools/j1939_84/controllers/part03/Part03Step15Controller.java

```diff
@@ -54,8 +60,29 @@
     @Override
     protected void run() throws Throwable {
         // 6.3.15.1.a. DS DM21 (send Request (PGN 59904) for PGN 49408 (SPNs 3069, 3295)) to each OBD ECU.
-        // 6.3.15.2.a. Fail if any ECU reports distance (SPN 3069) or time (SPN 3295) with MIL on > 0.
-        // 6.3.15.2.b. Fail if NACK not received from OBD ECUs that did not provide DM21 response to DS query.
+        getDataRepository()
+                .getObdModuleAddresses()
+                .stream()
+                .sorted()
+                .forEach(address -> getDiagnosticMessageModule().requestDM21(getListener(), address)
+                        .getPacket()
+                        .ifPresentOrElse(packet -> {
+                            if (packet.left.isPresent()) {
+                                DM21DiagnosticReadinessPacket dm21 = packet.left.get();
+                                if (dm21.getKmWhileMILIsActivated() > 0 || dm21.getMinutesWhileMILIsActivated() > 0) {
+                                    // 6.3.15.2.a. Fail if any ECU reports distance (SPN 3069) or time (SPN 3295) with MIL on > 0.
+                                    addFailure("6.3.15.2.a - OBD module " + dm21.getModuleName() + " reported active time or distance > 0");
+                                }
+                            } else if(packet.right.isPresent()){
+                                AcknowledgmentPacket ackPacket = packet.right.get();
+                                if(ackPacket.getResponse() != NACK){
+                                    addFailure("6.3.15.2.b - OBD module " + getAddressName(address) + " did not respond to DS DM21 request");
```

Comment:
Technically, the module _did_ response, but it wasn't a NACK.  The error (here and below) should be closer to the failure from the spec:

"6.3.15.2.b - NACK not received from <ModuleName> which did not provide a DM21 response to DS query"